error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/CollectionRules.java

```diff
@@ -184,6 +184,19 @@ void after(Set<T> removeFrom, Collection<S> elementsToRemove) {
     }
   }
 
+  /** Rely on Set's property to preserve distinct elements. */
```

Comment:
This does assume that the set doesn't use a custom equality, like a `SortedSet` might. Will add a note:
```suggestion
  /** Don't unnecessarily call {@link Stream#distinct()} on an already-unique stream of elements. */
  // XXX: This rule assumes that the `Set` relies on `Object#equals`, rather than a custom
  // equivalence relation.
  // XXX: Expressions that drop or reorder elements from the stream, such as `.filter`, `.skip` and
  // `sorted`, can similarly be simplified. Covering all cases is better done using an Error Prone
  // check.
```

(I'm not too worried about false positives: it's likely rare that there is code that wants cares about both types of deduplication in a single context.)